Pyic 8757 stop deployment on test failure

DanCorderIPV opened the pull request:
## Proposed changes
### What changed

Run automated tests after merging a PR as well as before

### Why did it change

To catch breaking changes that happen after a PR is raised

### Issue tracking
<!-- Jira ticket & other docs, like RFCs -->

- [PYIC-8757](https://govukverify.atlassian.net/browse/PYIC-8757)


[PYIC-8757]: https://govukverify.atlassian.net/browse/PYIC-8757?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

thebauSoftwire commented on .github/workflows/secure-post-merge.yml:
```diff
@@ -16,6 +16,25 @@ on:
+      packages: read
+      contents: read
+    secrets:
+      API_KEY_CIMIT: ${{ secrets.API_KEY_CIMIT }}
+      API_KEY_EVCS: ${{ secrets.API_KEY_EVCS }}
+      API_KEY_SIS: ${{ secrets.API_KEY_SIS }}
+      ASYNC_QUEUE_API_KEY: ${{ secrets.ASYNC_QUEUE_API_KEY }}
+      CRI_STUB_GEN_CRED_API_KEY: ${{ secrets.CRI_STUB_GEN_CRED_API_KEY }}
+      DCMAW_ASYNC_SECRET: ${{ secrets.DCMAW_ASYNC_SECRET }}
+      MANAGEMENT_CIMIT_STUB_API_KEY: ${{ secrets.MANAGEMENT_CIMIT_STUB_API_KEY }}
+      SONAR_TOKEN: ${{ secrets.SONAR_TOKEN }}
+      TICF_API_KEY: ${{ secrets.TICF_API_KEY }}
+      MANAGEMENT_TICF_API_KEY: ${{ secrets.MANAGEMENT_TICF_API_KEY }}
```
could we just use `secrets: inherit` here? for [passing secrets down to reusable workflows](https://docs.github.com/en/actions/how-tos/reuse-automations/reuse-workflows#passing-inputs-and-secrets-to-a-reusable-workflow)

DanCorderIPV replied:
SonarQube doesn't like that :(

thebauSoftwire replied:
:') ah fair enough

thebauSoftwire commented on .github/workflows/pre-merge-checks.yml:
```diff
@@ -15,155 +12,24 @@ env:
-          name: Application logs
-          path: api-tests/core-back-output.log
-          retention-days: 7
+    secrets:
+      API_KEY_CIMIT: ${{ secrets.API_KEY_CIMIT }}
+      API_KEY_EVCS: ${{ secrets.API_KEY_EVCS }}
+      API_KEY_SIS: ${{ secrets.API_KEY_SIS }}
+      ASYNC_QUEUE_API_KEY: ${{ secrets.ASYNC_QUEUE_API_KEY }}
+      CRI_STUB_GEN_CRED_API_KEY: ${{ secrets.CRI_STUB_GEN_CRED_API_KEY }}
+      DCMAW_ASYNC_SECRET: ${{ secrets.DCMAW_ASYNC_SECRET }}
+      MANAGEMENT_CIMIT_STUB_API_KEY: ${{ secrets.MANAGEMENT_CIMIT_STUB_API_KEY }}
+      SONAR_TOKEN: ${{ secrets.SONAR_TOKEN }}
+      TICF_API_KEY: ${{ secrets.TICF_API_KEY }}
+      MANAGEMENT_TICF_API_KEY: ${{ secrets.MANAGEMENT_TICF_API_KEY }}
```
same here (see [this comment](https://github.com/govuk-one-login/ipv-core-back/pull/3641/changes#r2631232376))

thebauSoftwire reviewed (dismissed):
Just one suggestion but looks good!

DanCorderIPV committed:
PYIC-8757: Run automated tests before deployment

Extract the automated test steps into a separate workflow file and then call that in both pre-merge and post-merge workflows to stop deployment if a conflicting change has been merged since the PR was raised.

DanCorderIPV committed:
PYIC-8757: Security improvements

Remove the chance of secrets being stored in command line logs or process lists
Explicitly stop npm scripts running (although this should be covered by the .npmrc file)

DanCorderIPV committed:
PYIC-8757: Improve job name

DanCorderIPV committed:
PYIC-8757: Use SHA rather than version for setup-sam action

undermad commented on .github/workflows/secure-post-merge.yml:
```diff
@@ -4,18 +4,28 @@ on:
   push:
     branches:
       - main
-    paths-ignore:
-      # If you are adding or removing from this list, also update 'files_ignore' in check-if-deploy-needed job
-      - '**/*.md'
-      - 'lambdas/*/src/test/**'
-      - 'libs/*/src/test/**'
-      - 'libs/test-data/**'
-      - 'libs/test-helpers/**'
-      - 'journey-map/**'
-      - 'local-running/**'
```
Hi Dan, I assume you removed it as pre-merge-check won't run on main merge and we want tests to be invoked event if deployment is not required? It looks like solid solution with simplified code.

DanCorderIPV replied:
Yes, also this means that the later steps in this file can depend on the tests and will not run if the tests fail.
The current way with pre-merge-checks can't stop the deployment if the tests fail as the two workflows run in parallel.

undermad reviewed (approved):
This is looking great, I just left a comment to verify that I understood correctly intention of deleted ignore paths. Also, I think deployment should be double checked as part of QA ticket to verify that API tests and lambdas are deployed or not depending on the changes.
